test/TensorFlowRuntime/tensor_group.swift

```diff
@@ -0,0 +1,78 @@
+// RUN: %empty-directory(%t)
+// RUN: %target-build-swift -swift-version 5 -g %s -o %t/a.out
+// RUN: %target-codesign %t/a.out
+// RUN: %target-run %t/a.out
+// REQUIRES: executable_test
+//
+// `TensorGroup` tests.
+
+import TensorFlow
+import StdlibUnittest
+
+var TensorGroupTests = TestSuite("TensorGroup")
+
+extension TensorDataType : Equatable {
+  public static func == (lhs: TensorDataType, rhs: TensorDataType) -> Bool {
+    return Int(lhs._cDataType.rawValue) == Int(rhs._cDataType.rawValue)
+  }
+}
+
+struct Empty : TensorGroup {}
+
+struct Simple : TensorGroup {
+  var w, b: Tensor<Float>
+}
+
+struct Mixed : TensorGroup {
+  // Mutable.
+  var string: StringTensor
+  var float: Tensor<Float>
+  // Immutable.
+  let int: Tensor<Int32>
+}
+
+struct Nested : TensorGroup {
+  // Immutable.
+  let simple: Simple
+  // Mutable.
+  var mixed: Mixed
+}
+
+struct Generic<T: TensorGroup, U: TensorGroup> : TensorGroup {
+  var t: T
+  var u: U
+}
+
+TensorGroupTests.test("Empty") {
+  expectEqual([], Empty._typeList)
+}
+
+TensorGroupTests.test("Simple") {
```

Comment:
Could you please add some tests for the other derived implementations too?
Namely `init(_owning:)` and `func _unpackTensorHandles(into address: UnsafeMutablePointer<CTensorHandle>?)`.

If derived implementations are correct, we can be confident they'll work with compiler infra for `TensorGroup` types.

Reply:
I wanted to test these implementations too, but I'm not sure how. How do I test for memory ownership of tensors? Do you guys have some standard way for doing that?

Reply:
@dan-zheng @rxwei I have addressed everything else except this because I'm not sure how I can test for memory ownership.

Reply:
I don't think you need to test ownership in particular - if you could please add tests for synthesized implementations of `init(_owning:)` and `func _unpackTensorHandles(into address: UnsafeMutablePointer<CTensorHandle>?)`, that would be great!

Here's an example for `init(_owning):` that passes with current hardcoded `TensorPair`:
```swift
import TensorFlow
let x = Tensor<Float>(1)
let y = Tensor<Float>(2)

let buffer = UnsafeMutableBufferPointer<CTensorHandle>.allocate(capacity: 2)
buffer.baseAddress?.initialize(to: x.handle._cTensorHandle)
buffer.baseAddress?.advanced(by: 1).initialize(to: y.handle._cTensorHandle)

// Initializer signature:
// init(_owning tensorHandles: UnsafePointer<CTensorHandle>?)
let pair = TensorPair<Tensor<Float>, Tensor<Float>>(
  _owning: UnsafePointer(buffer.baseAddress))
expectEqual(TensorPair(Tensor<Float>(1), Tensor<Float>(2)), pair)
```

If you haven't used them before, here are some resources for Swift pointer APIs:
* [`OpaquePointer`](https://developer.apple.com/documentation/swift/opaquepointer): this is the type of `Tensor(1).handle._cTensorHandle`. It's an opaque C pointer.
* [`UnsafePointer<T>`](https://developer.apple.com/documentation/swift/unsafepointer): this is a typed pointer.

Reply:
I added the tests for `_unpackTensorHandles`, but having issues with `init(_owning:)`. Actually, even the example you provide for the hardcoded `TensorPair` gives me a segfault. Are you sure that this example runs fine?

Reply:
I believe there may be a bug in the base cases for `TensorGroup` in the stdlib. I'm checking now and if indeed that's the case I'll add a fix in this PR.

Reply:
Even the following simple example fails with a segfault:

```swift
let x = Tensor<Float>(1)
let buffer = UnsafeMutableBufferPointer<CTensorHandle>.allocate(capacity: 1)
buffer.baseAddress?.initialize(to: x.handle._cTensorHandle)
let t = Tensor<Float>(_owning: UnsafePointer(buffer.baseAddress))
```

Reply:
> Actually, even the example you provide for the hardcoded `TensorPair` gives me a segfault. Are you sure that this example runs fine?

I ran my example using the [v0.2 macOS toolchain](https://github.com/tensorflow/swift/blob/master/Installation.md#pre-built-packages), sans `TensorGroup` derived conformances. It seems to work fine for me.

> Even the following simple example fails with a segfault:
That example runs fine for me too.

I wonder what build are you using to test these programs? Is it your development build for this PR?

Reply:
It is my development build with the changes from the master branch merged. I don't see how this should affect the single tensor example shown above though that does not use TensorGroups. Is there anything I should be aware of when putting the example code within a test as follows?

```swift
ExampleTests.test("SimpleInit") {
  let x = Tensor<Float>(1)
  let buffer = UnsafeMutableBufferPointer<CTensorHandle>.allocate(capacity: 1)
  buffer.baseAddress?.initialize(to: x.handle._cTensorHandle)
  let t = Tensor<Float>(_owning: UnsafePointer(buffer.baseAddress))
}
```

Reply:
Hmm. When I run the program as a standalone Swift file, it works fine:
```console
$ cat tensor.swift
import TensorFlow
let x = Tensor<Float>(1)
let buffer = UnsafeMutableBufferPointer<CTensorHandle>.allocate(capacity: 1)
buffer.baseAddress?.initialize(to: x.handle._cTensorHandle)
let t = Tensor<Float>(_owning: UnsafePointer(buffer.baseAddress))
print(t)
$ $SWIFT_BIN/swift tensor.swift
1.0
```

When I add it as a `TensorFlowRuntime` test and run with `lit` however, it does crash.
```
[ RUN      ] Tensor.SimpleInit
stderr>>> CRASHED: SIGSEGV
the test crashed unexpectedly
[     FAIL ] Tensor.SimpleInit
Tensor: Some tests failed, aborting
UXPASS: []
FAIL: ["SimpleInit"]
SKIP: []
```

Here's what lldb shows:
```
Process 93496 launched: '/Users/danielzheng/swift-dev/build/Ninja-ReleaseAssert+stdlib-Release/swift-macosx-x86_64/test-macosx-x86_64/TensorFlowRuntime/Output/simple.swift.tmp/a.out' (x86_64)
StdlibUnittest: using filter: SimpleInit
[ RUN      ] Tensor.SimpleInit
a.out(93496,0x1000c65c0) malloc: *** error for object 0x10e931050: pointer being freed was not allocated
a.out(93496,0x1000c65c0) malloc: *** set a breakpoint in malloc_error_break to debug
Process 93496 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x00007fff6be3623e libsystem_kernel.dylib`__pthread_kill + 10
libsystem_kernel.dylib`__pthread_kill:
->  0x7fff6be3623e <+10>: jae    0x7fff6be36248            ; <+20>
    0x7fff6be36240 <+12>: movq   %rax, %rdi
    0x7fff6be36243 <+15>: jmp    0x7fff6be303b7            ; cerror_nocancel
    0x7fff6be36248 <+20>: retq
Target 0: (a.out) stopped.
```

I don't quite have time to investigate this further sadly. My `RUN` line is `%target-run-eager-swift %swift-tensorflow-test-run-extra-options`, I wonder if this includes compiler flags that might affect compilation and cause `pointer being freed was not allocated`.

Reply:
Thanks for verifying this so I know it's not something in my setup. I believe there may be a more important underlying issue. In the above example, tensors `x` and `t` are both pointing to the same underlying memory managed by TF. In this case, when we deallocate `x`, for example, is that memory freed? If yes, then when deallocating `t` after that, we may get a segfault. Given the `_owning` argument label I would assume that `t` should own the memory, but then what happens with `x`? Because there is no way for `x` to know that `t` owns the underlying memory. Does this make sense?

Reply:
I see your point about the potential "underlying issue".
I wonder if someone more familiar with tensor packing/unpacking can comment on this: cc @marcrasi (who implemented it in https://github.com/apple/swift/pull/20003).
* What are your thoughts on the ownership issue raised by Anthony?
* Why is it that the test program above works when run as a standalone Swift program, but not as a TensorFlowRuntime lit test?

Reply:
This is wrong:
```
ExampleTests.test("SimpleInit") {
  let x = Tensor<Float>(1)
  let buffer = UnsafeMutableBufferPointer<CTensorHandle>.allocate(capacity: 1)
  buffer.baseAddress?.initialize(to: x.handle._cTensorHandle)
  let t = Tensor<Float>(_owning: UnsafePointer(buffer.baseAddress))
}
```
for exactly the reason @eaplatanios says: `x` and `t` both think that they own the underlying TF memory, so they both try to free it. I think the reason that it doesn't crash as standalone top-level code is that Swift might never run destructors for top-level variables.

To correctly call `init(_owning:)`, you'll need to manually construct some CTensorHandles that aren't owned by anything. Here's a way to do that (I haven't compiled or run this):
```
import CTensorFlow
ExampleTests.test("SimpleInit") {
  let x = Tensor<Float>(1)
  
  let status = TF_NewStatus()
  let cTensorHandle = TFE_TensorHandleCopySharingTensor(x.handle._cTensorHandle, status)
  TF_DeleteStatus(status)

  let buffer = UnsafeMutableBufferPointer<CTensorHandle>.allocate(capacity: 1)
  buffer.baseAddress?.initialize(to: cTensorHandle)
  let t = Tensor<Float>(_owning: UnsafePointer(buffer.baseAddress))
}
```

Hopefully all the existing places in S4TF that make calls to `init(_owning:)` do this correctly :). [Here's an example of a place](https://github.com/apple/swift/blob/0ad31669329e5825b14742b9de11648bc8a1e9eb/stdlib/public/TensorFlow/CompilerRuntime.swift#L791) that appears to do it correctly, so that's good.

Reply:
Thanks a lot @marcrasi ! That makes sense. :)

Reply:
Added the tests and they now all succeed! :)